File: the-fellowship-of-the-ring/the-fellowship-of-the-ring.go
package main

import (
	"fmt"
)

type token struct {
	emitter int
	receiver int
	message string
}

func worker(in, out chan token, id int) {
	resToken := <-in
	fmt.Println("Channel - ", id)
	close(in)
	if resToken.message == "DIE"{
		out <- resToken 
		return
	}
	if id == resToken.receiver {
			fmt.Println("I'm the receiver! The message i've got:", resToken.message)
			resToken.message = "DIE"
			if resToken.emitter-resToken.receiver > 1{
				out <- resToken 
			}
			return
	}else if resToken.emitter == id{
		fmt.Println("Reciever not found!")
		return
	} else{
		out <- resToken
	}
	
}

func main() {
	const chanQuan int = 10
	emitter := 2
	receiver:= 2
	message := "Some data"
	var chans [chanQuan]chan token 
	for i := range chans {
		chans[i] = make(chan token)
	}

	for i := 0; i < chanQuan-1; i++ {
		go worker(chans[i], chans[i+1], i)

	}
	go worker(chans[chanQuan-1], chans[0], 9)
	chans[emitter+1] <- token{emitter, receiver, message}
	<-chans[emitter]
}

File: the-fellowship-of-the-ring/the-fellowship-of-the-ring.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Threading;
using System.Collections.Concurrent;
namespace lab4
{
    public class Token
    {
        private string _message;
        private int _reciever;
        private int _emitter;

        public string Message { get { return _message; } set { _message = value; } }
        public int Reciever { get { return _reciever; } }
        public int Emitter { get { return _emitter; } }


        public Token(string message, int reciever, int emitter)
        {
            _message = message;
            _reciever = reciever;
            _emitter = emitter;
        }
    }

    class Program
    {
        const int ThreadQuan = 10;
        const int emitter = 7;
        const int reciever = 30;
        const string message = "Message";
        static void Main(string[] args)
        {
            var conQueues = new ConcurrentQueue<Token>[ThreadQuan];
            for (var i = 0; i < ThreadQuan; i++)
            {
                conQueues[i] = new ConcurrentQueue<Token>();
            }
            var threads = new Thread[ThreadQuan];
            for (var i = 0; i < ThreadQuan - 1; i++)
            {
                var index = i;
                threads[i] = new Thread(() =>
                {
                    NewThread(conQueues[index], conQueues[index + 1], index);
                });
            }
            threads[ThreadQuan - 1] = new Thread(() =>
            {
                NewThread(conQueues[ThreadQuan - 1], conQueues[0], ThreadQuan - 1);
            });
            for (var i = 0; i < ThreadQuan; i++)
            {
                threads[i].Start();
            }
            Token token = new Token(message, reciever, emitter);
            conQueues[emitter+1].Enqueue(token);
            threads[emitter].Join();
            Console.Read();
        }

        unsafe static void NewThread(ConcurrentQueue<Token> inQueue, ConcurrentQueue<Token> outQueue, int id)
        {
            while (true)
            {
                Token resToken;
                if (!inQueue.TryDequeue(out resToken))
                {
                    Thread.Sleep(100);
                    continue;
                }
                else
                {
                    if(resToken.Message == "DIE")
                    {
                        outQueue.Enqueue(resToken);
                        break;
                    }
                    if (resToken.Emitter == id)
                    {
                        Console.WriteLine("Reciever not found!");
                        break;
                    }
                    else if (resToken.Reciever == id)
                    {
                        Console.WriteLine(string.Format("I'm the receiver! The message i've got:{0}", resToken.Message));
                        resToken.Message = "DIE";
                        outQueue.Enqueue(resToken);
                        break;
                    }
                    else
                    {
                        outQueue.Enqueue(resToken);
                    }
                }
                
            }
        }
    }
}
